solana/indexer/dbc/indexer.go

```diff
@@ -0,0 +1,305 @@
+		logger:           logger.Named("DBCIndexer"),
+	}
+}
+
+func (d *Indexer) Start(ctx context.Context) {
+	// To ensure only one subscription task is running at a time, keep track of
+	// the last cancel function and call it on the next notification.
+	var lastCancel context.CancelFunc
+
+	// Ensure all gRPC clients are closed on shutdown
+	var grpcClients []common.GrpcClient
+	defer (func() {
+		for _, client := range grpcClients {
+			client.Close()
+		}
+	})()
+
+	// On notification, cancel the previous subscription task (if any) and start a new one
+	handleNotif := func(ctx context.Context, notification *pgconn.Notification) {
+		subCtx, cancel := context.WithCancel(ctx)
+
+		// Cancel previous subscription task
+		if lastCancel != nil {
+			lastCancel()
+		}
+
+		// Close previous gRPC clients
+		for _, client := range grpcClients {
+			client.Close()
+		}
+
+		// Resubscribe to all DBC pools
+		// TODO: Optimize this to only add/remove DBC pools instead of resubscribing to all
+		clients, err := d.subscribe(subCtx)
+		grpcClients = clients
+		if err != nil {
+			d.logger.Error("failed to resubscribe to DBC pools", zap.Error(err))
+			cancel()
+			return
+		}
+
+		lastCancel = cancel
+	}
+
+	// Setup initial subscription
+	clients, err := d.subscribe(ctx)
+	if err != nil {
+		d.logger.Error("failed to subscribe to DAMM V2 pools", zap.Error(err))
+		return
+	}
+	grpcClients = clients
+
+	// Watch for new pools to be added
+	err = common.WatchPgNotification(ctx, d.pool, NOTIFICATION_NAME, handleNotif, d.logger)
+	if err != nil {
+		d.logger.Error("failed to watch for DBC pool changes", zap.Error(err))
+		return
+	}
+
+	// Wait for shutdown
+	for {
+		select {
+		case <-ctx.Done():
+			d.logger.Info("received shutdown signal, stopping indexer")
+			return
+		default:
+		}
+	}
+}
+
+func (d *Indexer) HandleUpdate(ctx context.Context, msg *pb.SubscribeUpdate) error {
+	// Handle slot updates
+	slotUpdate := msg.GetSlot()
+	if slotUpdate != nil {
+		// only update every 10 slots to reduce db load and write latency
+		if slotUpdate.Slot%10 == 0 {
+			// Use the filter as the checkpoint ID
+			checkpointId := msg.Filters[0]
+
+			err := common.UpdateCheckpoint(ctx, d.pool, checkpointId, slotUpdate.Slot)
+			if err != nil {
+				d.logger.Error("failed to update slot checkpoint", zap.Error(err))
+			}
+		}
+	}
+
+	// Handle account updates
+	accountUpdate := msg.GetAccount()
+	if accountUpdate != nil {
+		var pool meteora_dbc.Pool
+		err := bin.NewBorshDecoder(accountUpdate.Account.Data).Decode(&pool)
+		if err != nil {
+			return fmt.Errorf("failed to decode DBC pool account: %w", err)
+		}
+
+		account := solana.PublicKeyFromBytes(accountUpdate.Account.Pubkey)
+		err = upsertDbcPool(ctx, d.pool, accountUpdate.Slot, account, &pool)
+		if err != nil {
+			return fmt.Errorf("failed to upsert DBC pool: %w", err)
+		}
+
+		// If the pool is migrated, check for the migration transaction and process it
+		if pool.IsMigrated == uint8(1) {
```

Comment:
Is this safe to run over and over after it's migrated? I imagine it's possible to get transactions against the dbc pool after it's already been migrated.

Reply:
Safe, yes. Useful, no. I imagine the DBC account is pretty untouched after migration anyway though